src/frontend/apps/impress/src/features/docs/doc-management/libs/CollaborationProvider.ts

```diff
@@ -0,0 +1,333 @@
+   * We use this variable to avoid treating the same awareness message twice.
+   */
+  private treatedAwarenessMessage: string | null = null;
+  /**
+   * Server-Sent Events
+   */
+  protected sse: EventSource | null = null;
+  /**
+   * Polling timeout
+   * It is used to avoid starting the polling to quickly
+   * to let the class init properly.
+   */
+  protected pollTimeout: NodeJS.Timeout | null = null;
+  /**
+   * Easy way to get the url of the server
+   */
+  protected url = '';
+
+  public constructor(configuration: CollaborationProviderConfiguration) {
+    let url = '';
+    if (isHocuspocusProviderConfigurationUrl(configuration)) {
+      url = configuration.url;
+      let withWS = true;
+      if (
+        new URLSearchParams(window.location.search).get('withoutWS') === 'true'
+      ) {
+        withWS = false;
+      }
+
+      configuration.url = !withWS ? 'ws://localhost:6666' : configuration.url;
+    }
+
+    super(configuration);
+
+    this.url = url;
+    this.canEdit = configuration.canEdit;
+
+    if (configuration.canEdit) {
+      this.on('outgoingMessage', this.onPollOutgoingMessage.bind(this));
+    }
+  }
+
+  public setPollDefaultValues(): void {
+    this.isLongPollingStarted = false;
+    this.isWebsocketFailed = false;
+    this.seemsUnsyncCount = 0;
+    this.sse?.close();
+    this.sse = null;
+    if (this.pollTimeout) {
+      clearTimeout(this.pollTimeout);
+    }
+  }
+
+  public destroy(): void {
+    super.destroy();
+    this.setPollDefaultValues();
+  }
+
+  public onStatus({ status }: onStatusParameters) {
+    if (status === WebSocketStatus.Connecting) {
+      this.isWebsocketFailed = true;
+      if (this.pollTimeout) {
+        clearTimeout(this.pollTimeout);
+      }
+      this.pollTimeout = setTimeout(() => {
+        this.initPolling();
+      }, 5000);
+    } else if (status === WebSocketStatus.Connected) {
+      this.setPollDefaultValues();
+    }
+
+    super.onStatus({ status });
+  }
+
+  public initPolling() {
+    if (this.isLongPollingStarted || !this.isWebsocketFailed) {
+      return;
+    }
+
+    this.isLongPollingStarted = true;
+    void this.pollSync(true);
+    this.initCollaborationSSE();
+  }
+
+  protected toPollUrl(endpoint: string): string {
+    let pollUrl = this.url.replace('ws:', 'http:');
+    if (pollUrl.includes('wss:')) {
+      pollUrl = pollUrl.replace('wss:', 'https:');
+    }
+
+    pollUrl = pollUrl.replace('/ws/', '/ws/poll/' + endpoint + '/');
+
+    // To have our requests not cached
+    return `${pollUrl}&${Date.now()}`;
+  }
+
+  protected isDuplicateAwareness(message64: string): boolean {
+    if (this.treatedAwarenessMessage === message64) {
+      return true;
+    }
+
+    this.treatedAwarenessMessage = message64;
+
+    return false;
+  }
+
+  /**
+   * Outgoing message event
+   *
+   * Sent to the server the message to
+   * be sent to the other users
+   */
+  public async onPollOutgoingMessage({ message }: onOutgoingMessageParameters) {
```

Comment:
I got confused at some parts when checking the code, I think it might be helpful to explain or improve the naming a little bit.

Technically this method is not "polling" anymore, right? When you're polling, you retrieve an update from somewhere, but here we're just sending a message, like a regular POST request.

---

Also, you're not using Long-Polling at all, because long-polling means (afaik) that you send a request to the server, which keeps the connection open until the server has something interesting to send to you. I don't think this is the case, as you're using SSE for events from server -> client.

Which is fine, but maybe better to avoid the term long-polling then to avoid confusion down the line